### PR TITLE
RavenDB_15750 Adding tests and fixes into RQL Antlr gramma.

### PR DESCRIPTION
**Pull request by maciejaszyk**

### CODE REVIEW

File: src/Raven.Studio/languageService/grammar/RqlLexer.g4 [hunk 1]
```diff
@@ -74,11 +75,22 @@ JAVASCRIPT: '{' ( JAVASCRIPT | ~'{'  | ~'}' )*? '}'   -> channel(2);
 //Literals
 JS_FUNCTION_DECLARATION: 'declare function';
 NUM: DIGIT+ (DOT DIGIT+)?;
-STRING : ('"' ( '\\"' | . )*? '"' ) | ('\'' ( '\\"' | . )*? '\'' ) ;
-WORD: AT? [a-zA-Z_0-9]+;
+//STRING : ('"' ( '\\"' | . )*? '"' ) | ('\'' ( '\\"' | . )*? '\'' ) ;
```
Comment (ml054) [inline src/Raven.Studio/languageService/grammar/RqlLexer.g4:78]: remove?

File: src/Raven.Studio/languageService/grammar/RqlParser.g4 [hunk 2]
```diff
@@ -137,54 +118,64 @@ groupByStatement:
...
-    | NUM
-    | (TRUE | FALSE)
-    | function
-    | variable
+    left=expr binary right=expr #binaryExpression
+    | OP_PAR expr CL_PAR #oppar
+    |left=exprValue EQUAL right=exprValue #equalExpression
+    |left=exprValue MATH right=exprValue #mathExpression
+    |specialFunctions #specialFunctionst
+    | inFuction #inExpr
```
Comment (ml054) [inline src/Raven.Studio/languageService/grammar/RqlParser.g4:142]: fix ident

File: src/Raven.Studio/languageService/grammar/RqlParser.g4 [hunk 3]
```diff
@@ -137,54 +118,64 @@ groupByStatement:
...
-    | expr AND NOT? expr
-    | expr OR NOT? expr
-    | ID OP_PAR CL_PAR
-    | STRING
-    | NUM
-    | (TRUE | FALSE)
-    | function
-    | variable
+    left=expr binary right=expr #binaryExpression
+    | OP_PAR expr CL_PAR #oppar
```
Comment (ml054) [inline src/Raven.Studio/languageService/grammar/RqlParser.g4:138]: #op_par ?

Comment (ml054) [reply src/Raven.Studio/languageService/grammar/RqlParser.g4:138]: or orPar to match casing

File: src/Raven.Studio/languageService/test/grammar/fromStatement.spec.ts [hunk 4]
```diff
@@ -47,7 +47,43 @@ describe("FROM statement parser", function() {
...
+    it("from index '", function() {
+        const { parseTree, parser } = parseRql(`from index '`);
+
+        expect(parser.numberOfSyntaxErrors)
+            .toEqual(1);
+
+        const from = parseTree.fromStatement();
+
+        expect(from)
+        .toBeInstanceOf(CollectionByIndexContext);
```
Comment (ml054) [inline src/Raven.Studio/languageService/test/grammar/fromStatement.spec.ts:75]: ident

File: src/Raven.Studio/languageService/test/grammar/fromStatement.spec.ts [hunk 5]
```diff
@@ -47,7 +47,43 @@ describe("FROM statement parser", function() {
...
+    it("from index d", function() {
+        const { parseTree, parser } = parseRql(`from index d"`);
+
+        expect(parser.numberOfSyntaxErrors)
+            .toEqual(1);
+
+        const from = parseTree.fromStatement();
+
+        expect(from)
+            .toBeInstanceOf(CollectionByIndexContext);
```
Comment (ml054) [inline src/Raven.Studio/languageService/test/grammar/fromStatement.spec.ts:87]: assert on index name

File: src/Raven.Studio/languageService/test/grammar/groupByStatement.spec.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
+﻿import {parseRql} from "../../src/parser";
```
Comment (ml054) [inline src/Raven.Studio/languageService/test/grammar/groupByStatement.spec.ts:1]: spaces around `{` `}` + fix Rider setup to automatically insert spaces in imports

File: src/Raven.Studio/languageService/test/grammar/groupByStatement.spec.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import {
+    GroupByStatementContext,
+    ParameterWithOptionalAliasContext,
+} from "../../src/generated/RqlParser";
+
+describe("GROUP BY statement parser", function () {
```
Comment (ml054) [inline src/Raven.Studio/languageService/test/grammar/groupByStatement.spec.ts:7]: remove space after `function` - in all places

File: src/Raven.Studio/languageService/test/grammar/groupByStatement.spec.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    it("single", function () {
+        const {parseTree, parser} = parseRql("from test group by zebra");
```
Comment (ml054) [inline src/Raven.Studio/languageService/test/grammar/groupByStatement.spec.ts:9]: spaces around `{`

File: src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        expect(parser.numberOfSyntaxErrors)
+            .toEqual(0);
+
+        const load = parseTree.loadStatement();
+        expect(load)
+            .toBeInstanceOf(LoadStatementContext);
+        expect(load._item._name.text)
+            .toEqual("x");
+        
+        expect(load._item._as.text).toEqual("asy");
```
Comment (ml054) [inline src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts:20]: can we assert y not asy here?

File: src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    });
+
+    it("array", function () {
+        const {parseTree, parser} = parseRql("from test load t as t, t as t");
```
Comment (ml054) [inline src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts:24]: can we avoid duplicates here? a as mappedA, b as mappedB

File: src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+        expect(parser.numberOfSyntaxErrors)
+            .toEqual(0);
+        const load = parseTree.loadStatement();
+        expect(load)
+            .toBeInstanceOf(LoadStatementContext);
+        // expect(load).con
```
Comment (ml054) [inline src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts:31]: remove?

File: src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        for (let index = 0; index < load.children.length; index++) {
```
Comment (ml054) [inline src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts:32]: avoid for loop

File: src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        }
+            
+        // for(i = 0; i < load.childCount; ++i)
+        // {
+        //    
+        // }
+        // expect(load._item._name.text)
+        //     .toEqual("x");
+        //
+        // expect(load._item._as.text).toEqual("asy");
```
Comment (ml054) [inline src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts:52]: remove?

File: src/Raven.Studio/languageService/test/grammar/orderBy.spec.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+        const orderBy = parseTree.orderByStatement();
+        
+        expect(orderBy)
+            .toBeInstanceOf(OrderByStatementContext);
+        expect(orderBy._value)
+            .toBeInstanceOf(OrderByItemContext);
+        
+        const item = orderBy._value;
+        expect(item._value.text).toEqual("item")
+        expect(item._order.text).toEqual("asALPHANUMERIC");
```
Comment (ml054) [inline src/Raven.Studio/languageService/test/grammar/orderBy.spec.ts:23]: avoid such text to compare

File: src/Raven.Studio/languageService/test/grammar/select.spec.ts [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+        expect(parser.numberOfSyntaxErrors)
+            .toEqual(0);
+
+        const select = parseTree.selectStatement();
+        
+        expect(select)
+            .toBeInstanceOf(ProjectIndividualFieldsContext);
+        expect(select.text)
+            .toEqual("selectx");
```
Comment (ml054) [inline src/Raven.Studio/languageService/test/grammar/select.spec.ts:20]: avoid such expect?

File: src/Raven.Studio/languageService/test/grammar/select.spec.ts [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    it("distinct", function () {
+        const {parseTree, parser} = parseRql("from test order by item as ALPHANUMERIC desc select distinct *");
+
+        expect(parser.numberOfSyntaxErrors)
+            .toEqual(0);
+
+        const select = parseTree.selectStatement();
+
+        expect(select)
+            .toBeInstanceOf(GetAllDistinctContext);
```
Comment (ml054) [inline src/Raven.Studio/languageService/test/grammar/select.spec.ts:31]: any assert on `*`?

File: src/Raven.Studio/languageService/test/grammar/select.spec.ts [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+
+    it("javascript", function () {
+        const {parseTree, parser} = parseRql("from test order by item as ALPHANUMERIC desc select {test;}");
+
+        expect(parser.numberOfSyntaxErrors)
+            .toEqual(0);
+
+        const select = parseTree.selectStatement();
+
+        expect(select)
```
Comment (ml054) [inline src/Raven.Studio/languageService/test/grammar/select.spec.ts:42]: can we assert on captured js text?

File: src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts (hunk 9 above)
Comment (maciejaszyk) [reply src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts:20]: Yes, we can go deeper in AST to AliasTree

File: src/Raven.Studio/languageService/grammar/RqlLexer.g4 (hunk 1 above)
Comment (maciejaszyk) [reply src/Raven.Studio/languageService/grammar/RqlLexer.g4:78]: done

File: src/Raven.Studio/languageService/grammar/RqlParser.g4 (hunk 3 above)
Comment (maciejaszyk) [reply src/Raven.Studio/languageService/grammar/RqlParser.g4:138]: done

File: src/Raven.Studio/languageService/test/grammar/fromStatement.spec.ts (hunk 4 above)
Comment (maciejaszyk) [reply src/Raven.Studio/languageService/test/grammar/fromStatement.spec.ts:75]: done

File: src/Raven.Studio/languageService/test/grammar/fromStatement.spec.ts (hunk 5 above)
Comment (maciejaszyk) [reply src/Raven.Studio/languageService/test/grammar/fromStatement.spec.ts:87]: done

File: src/Raven.Studio/languageService/test/grammar/groupByStatement.spec.ts (hunk 6 above)
Comment (maciejaszyk) [reply src/Raven.Studio/languageService/test/grammar/groupByStatement.spec.ts:1]: done

File: src/Raven.Studio/languageService/test/grammar/groupByStatement.spec.ts (hunk 7 above)
Comment (maciejaszyk) [reply src/Raven.Studio/languageService/test/grammar/groupByStatement.spec.ts:7]: done

File: src/Raven.Studio/languageService/test/grammar/groupByStatement.spec.ts (hunk 8 above)
Comment (maciejaszyk) [reply src/Raven.Studio/languageService/test/grammar/groupByStatement.spec.ts:9]: done

File: src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts (hunk 9 above)
Comment (maciejaszyk) [reply src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts:20]: done

File: src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts (hunk 10 above)
Comment (maciejaszyk) [reply src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts:24]: Duplicates a lot of code and changed nothing, we assert subtrees.

File: src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts (hunk 11 above)
Comment (maciejaszyk) [reply src/Raven.Studio/languageService/test/grammar/loadStatement.spec.ts:31]: done

File: src/Raven.Studio/languageService/test/grammar/orderBy.spec.ts (hunk 14 above)
Comment (maciejaszyk) [reply src/Raven.Studio/languageService/test/grammar/orderBy.spec.ts:23]: ?
Changed into:
image

File: src/Raven.Studio/languageService/test/grammar/select.spec.ts (hunk 16 above)
Comment (maciejaszyk) [reply src/Raven.Studio/languageService/test/grammar/select.spec.ts:31]: Don't need it.

File: src/Raven.Studio/languageService/test/grammar/select.spec.ts (hunk 17 above)
Comment (maciejaszyk) [reply src/Raven.Studio/languageService/test/grammar/select.spec.ts:42]: Probably yes, Idk how to do it in ts.

**Pull request merged**